kinetis: Make hwtimer_arch more configurable.

gebart opened the pull request:
This was a necessary change to support older K60 rev 1.x silicon which has a slightly different LPTMR module (The CNR register lacks the latching functionality of the current LPTMR)

Added some configuration macros to hwtimer module:
 - Added LPTIMER_DEV for choosing LPTMR module.
 - Added LPTIMER_CNR_NEEDS_LATCHING for defining whether the CPU
   requires latching of the CNR variable before reading its value.
 - Added LPTIMER_CLK{EN,DIS}() for enabling/disabling the clock gate
   to the LPTMR module.
 - Added LPTIMER_IRQ_CHAN for NVIC IRQ channel configuration.
 - Moved LPTIMER_IRQ_PRIO to cpu-conf.h

Additions necessary to cpu-conf.h:

    #define LPTIMER_DEV                      (LPTMR0) /**< LPTIMER hardware module */
    #define LPTIMER_CLKEN()                  (BITBAND_REG(SIM->SCGC5, SIM_SCGC5_LPTIMER_SHIFT) = 1)    /**< Enable LPTMR0 clock gate */
    #define LPTIMER_CLKDIS()                 (BITBAND_REG(SIM->SCGC5, SIM_SCGC5_LPTIMER_SHIFT) = 0)    /**< Disable LPTMR0 clock gate */
    /** IRQ priority for hwtimer interrupts */
    #define LPTIMER_IRQ_PRIO          1
    /** IRQ channel for hwtimer interrupts */
    #define LPTIMER_IRQ_CHAN          LPTMR0_IRQn
    #define LPTIMER_CNR_NEEDS_LATCHING 1

Signed-off-by: Joakim Gebart <joakim.gebart@eistec.se>

jfischer-phytec-iot commented:
As I see you using BITBAND_REG macro, why is it better or easier to understand than
SIM->SCGC5 |= SIM_SCGC5_LPTIMER_MASK;
or
SIM->SCGC5 &= ~SIM_SCGC5_LPTIMER_MASK;
?

gebart commented:
There's no difference, really. I have gotten into habit of writing `BITBAND_REG` whenever changing single bits.

jfischer-phytec-iot commented:
BITBAND_REG macro is Freescale specific and is not used by any other at RIOT.
Is it ok for you, do not use BITBAND_REG?

gebart commented:
@jfischer-phytec-iot Sometimes BITBAND_REG saves about 4 bytes of rom, but there is no big difference when it comes to these initialization functions, it is more useful when you can save clock cycles on repeated read-modify-write cycles. Also, I believe it ensures RMW atomicity by using the bitband region.

jfischer-phytec-iot commented:
@gebart Ok, so it uses Bit Manipulation Engine (BME)?
Then all kinetis driver should use it?

gebart commented:
@jfischer-phytec-iot The K60 uses bit banding, the KL devices have the BME which is supposedly more powerful than just bit banding.
https://community.freescale.com/thread/309727 

See also: 
 - MKW2x reference manual chapter 11.2.1 Aliased bit-band regions
 - MK60D10 reference manual chapter 4.2.1 Aliased bit-band regions (same text as MKW2x ref man)

gebart commented:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0439b/Behcjiic.html contains a description of this optional Cortex-M3+M4 feature

gebart commented:
Addressed @jfischer-phytec-iot comments (https://github.com/gebart/RIOT/commit/52a14de12dd12071f81400942c7e7cc3cef938fd#commitcomment-9241500)

Introduced internal lptmr_get_cnr() for handling both types of LPTMR modules (with and without latching)
